### PR TITLE
330 [Feature Request]: Option to not show module_add

### PR DESCRIPTION
**Pull request by kartikeyakirar**
this fixes https://github.com/insightsengineering/teal.slice/issues/330

In this pull request (PR), the module_add parameter is introduced in teal.slice::filter_settings(). This addition conceals the module that handles the display of dropdowns containing variables used to add a filter from filteredData.R.

### testing module
```
funny_module <- function (label = "Filter states", datanames = "all") {
  checkmate::assert_string(label)
  module(
    label = label,
    filters = datanames,
    ui = function(id, ...) {
      ns <- NS(id)
      div(
        h2("The following filter calls are generated:"),
        verbatimTextOutput(ns("filter_states")),
        verbatimTextOutput(ns("filter_calls")),
        actionButton(ns("reset"), "reset_to_default")
      )
    },
    server = function(input, output, session, data, filter_panel_api) {
      checkmate::assert_class(data, "tdata")
      observeEvent(input$reset, set_filter_state(filter_panel_api, default_filters))
      output$filter_states <-  renderPrint({
        logger::log_trace("rendering text1")
        filter_panel_api %>% get_filter_state()
      })
      output$filter_calls <- renderText({
        logger::log_trace("rendering text2")
        attr(data, "code")()
      })
    }
  )
}
```

### CDISC example
```
options(teal.log_level = "TRACE", teal.show_js_log = TRUE)
#options("teal.bs_theme" = bslib::bs_theme(version = 4))
#options(shiny.trace = TRUE)
devtools::load_all("teal.slice")
library(teal)
library(scda)

ADSL <- synthetic_cdisc_data("latest")$adsl
ADSL$empty <- NA
ADSL$logical1 <- FALSE
ADSL$logical <- sample(c(TRUE, FALSE), size = nrow(ADSL), replace = TRUE)
ADSL$numeric <- rnorm(nrow(ADSL))
ADSL$categorical2 <- sample(letters[1:10], size = nrow(ADSL), replace = TRUE)
ADSL$categorical <- sample(letters[1:3], size = nrow(ADSL), replace = TRUE, prob = c(.1, .3, .6))
ADSL$date <- Sys.Date() + seq_len(nrow(ADSL))
ADSL$date2 <- rep(Sys.Date() + 1:3, length.out = nrow(ADSL))
ADSL$datetime <- Sys.time() + seq_len(nrow(ADSL)) * 3600 * 12
ADSL$datetime2 <- rep(Sys.time() + 1:3 * 43200, length.out = nrow(ADSL))

ADSL$numeric[sample(1:nrow(ADSL), size = 10)] <- NA
ADSL$numeric[sample(1:nrow(ADSL), size = 10)] <- Inf
ADSL$logical[sample(1:nrow(ADSL), size = 10)] <- NA
ADSL$date[sample(1:nrow(ADSL), size = 10)] <- NA
ADSL$datetime[sample(1:nrow(ADSL), size = 10)] <- NA
ADSL$categorical2[sample(1:nrow(ADSL), size = 10)] <- NA
ADSL$categorical[sample(1:nrow(ADSL), size = 10)] <- NA

ADTTE <- synthetic_cdisc_data("latest")$adtte
ADRS <- synthetic_cdisc_data("latest")$adrs

ADTTE$numeric <- rnorm(nrow(ADTTE))
ADTTE$numeric[sample(1:nrow(ADTTE), size = 10,)] <- NA

default_filters <- filter_settings(
  filter_var(dataname = "ADSL", varname = "categorical", selected = c("a", "b"), disabled = TRUE),
  filter_var(dataname = "ADSL", varname = "categorical2", selected = c("a", "b"), disabled = TRUE),
  filter_var(dataname = "ADSL", varname = "numeric", selected = c(0, 140), keep_na = TRUE, keep_inf = TRUE, disabled = TRUE),
  filter_var(dataname = "ADSL", varname = "logical", selected = c(T), keep_na = TRUE, keep_inf = TRUE, disabled = TRUE),
  filter_var(dataname = "ADSL", varname = "datetime", disabled = TRUE),
  filter_var(dataname = "ADSL", varname = "date2", disabled = TRUE),
  filter_expr(id = "AF", title = "ADULT FEMALE", dataname = "ADSL", expr = "SEX %in% 'F' & AGE >= 18L", disabled = TRUE),
  filter_expr(id = "SE", title = "Safety-Evaluable", dataname = "ADSL", expr = "SAFFL == 'Y'", disabled = TRUE),
  filter_var(dataname = "ADSL", varname = "COUNTRY", selected = c("USA", "CAN", "JPN"), fixed = T, disabled = TRUE),
  count_type = "all",
  #include_varnames = list(ADSL = c("SEX", "categorical2", "numeric", "logical", "date", "datetime", "date2", "datetime2", "COUNTRY")),
  exclude_varnames = list(
    ADTTE = intersect(colnames(ADSL), colnames(ADTTE)),
    ADRS = colnames(ADSL)
  ),
  module_add = FALSE
)

data <- cdisc_data(
  cdisc_dataset("ADSL", ADSL),
  cdisc_dataset("ADTTE", ADTTE),
  cdisc_dataset("ADRS", ADRS)
)

app <- init(
  data = data,
  modules = list(
    tm_data_table(
      variables_selected = list(ADSL = c("STUDYID", "USUBJID", "SUBJID", "SITEID", "AGE", "SEX")),
      dt_args = list(caption = "ADSL Table Caption")
    ),
    funny_module()
  ),
  filter = default_filters
)

runApp(app)

```

### CODE REVIEW

**Comment by kartikeyakirar** [comment]
The module_add function is insufficiently descriptive and lacks clarity regarding its purpose.  It conceals a module responsible for managing the display of dropdown menus that contain variables utilized for adding filters.

**@chlebowa** @m7pr  Do you have any suggestion for naming this parameter ?

The visibility of the marked filter is determined by the value of the module_add flag.
image

File: R/FilterStates.R [hunk 1]
```diff
@@ -240,6 +240,10 @@ FilterStates <- R6::R6Class( # nolint
         )
       }
 
+      if (!is.null(private$module_add)) {
+        attr(fs, "module_add") <- private$module_add
+      }
```
Comment (gogonzo) [inline R/FilterStates.R:245]: I think that attribute can be retained in the `FilteredData$get_filter_state` instead. We don't need to maintaing `private$module_add` in both classes.

File: R/FilterStates.R [hunk 2]
```diff
@@ -273,6 +277,10 @@ FilterStates <- R6::R6Class( # nolint
         if (length(count_type)) {
           private$count_type <- count_type
         }
+        module_add <- attr(state, "module_add")
+        if (!is.null(module_add)) {
+          private$module_add <- module_add
+        }
```
Comment (gogonzo) [inline R/FilterStates.R:283]: Same as above
```suggestion
```

File: R/teal_slice.R [hunk 3]
```diff
@@ -73,21 +72,17 @@
 #'  match names of data sets and vector elements match variable names in respective data sets;
 #'  specify which variables are allowed to be filtered; see `Details`
 #' @param count_type `character(1)` string specifying how observations are tallied by these filter states.
+#' @param module_add `logical(1)` logical flag specifying whether to allow user to add filters (forbid filter addition).
```
Comment (gogonzo) [inline R/teal_slice.R:75]:
```suggestion
#' @param module_add `logical(1)` logical flag specifying whether to allow user to create new filters from available columns (forbid filter addition).
```

**Review by gogonzo — COMMENTED** [review]
`FilteredData$get_filter_state` should retain `private$module_add` and set it back again to a returned `teal_slices` object

File: R/teal_slice.R [hunk 4]
```diff
@@ -72,6 +72,7 @@
 #'  match names of data sets and vector elements match variable names in respective data sets;
 #'  specify which variables are allowed to be filtered; see `Details`
 #' @param count_type `character(1)` string specifying how observations are tallied by these filter states.
+#' @param module_add `logical(1)` logical flag specifying whether to allow user to create new filters from available columns (forbid filter addition).
```
Comment (chlebowa) [inline R/teal_slice.R:75]:
```suggestion
#' @param module_add `logical(1)` logical flag specifying whether the user will be able to add new filters
```

**Review by gogonzo — APPROVED** [review]
👍

**Comment by kartikeyakirar** [comment]
Adding refactoring parameter name as separate issue. it can be good first issue as well .https://github.com/insightsengineering/teal.slice/issues/355

**Pull request merged**